Add collection manage plugin

sean-m-sullivan opened the pull request:
### What does this PR do?
Add a collection manage plugin
        - If present will return data on a collection.
        - If present with version, will return data on a collection version.
        - If absent without version, will delete the collection and all versions.
        - If absent with version, will delete only specified version.

### How should this be tested?
Automated tests

I wonder if I should merge in the functionality of ah_collection_upload to this, Upload if specified path, then get its data, and delete if state is absent.

sean-m-sullivan committed:
add management of collection

sean-m-sullivan committed:
Merge branch 'redhat-cop:devel' into devel

sean-m-sullivan committed:
Merge branch 'redhat-cop:devel' into devel

sean-m-sullivan committed:
Merge branch 'devel' of github.com:sean-m-sullivan/ah_configuration into devel

Tompage1994 commented on plugins/modules/ah_collection_manage.py:
```diff
@@ -0,0 +1,115 @@
+
+DOCUMENTATION = """
+---
+module: ah_collection_manage
```
Not sure `_manage` fits in with usual naming for a module. This should potentially just be `ah_collection`

sean-m-sullivan replied:
With the addition of the upload yes, I was trying to make it so it was differentiated.

sean-m-sullivan replied:
Also thinking its going to be good to return the data anyways once its created, for reference, The data part, was more to have Something returned, why not get the collection data.

Tompage1994 reviewed (changes requested):
I like this. I haven't had a chance to test but will do before merging. 

I do think we should add the functionality of collection_upload into this and simply have one `ah_collection` module which does both things. I don't think it entirely makes sense to have `state: present` not actually cause an action (other than return data) since that doesn't fit in with creating a desired state. I would essentially just take the `ah_collection_upload` module and add the ability to delete collections. In other words `path` will probably be a required option when state == present

sean-m-sullivan committed:
update collection upload

sean-m-sullivan committed:
update collection upload

sean-m-sullivan committed:
updated lint

sean-m-sullivan committed:
updated lint

sean-m-sullivan committed:
update lint

sean-m-sullivan commented:
updated to make the changes, integrate publish, and make it so it does best of both

Tompage1994 commented on plugins/modules/ah_collection.py:
```diff
@@ -0,0 +1,135 @@
+
+EXAMPLES = """
+- name: Upload collection to automation hub
+  ah_collection_upload:
```
Should we not just remove the collection upload module completely?

Tompage1994 reviewed (changes requested):
This flow doesn't quite work if you specify the version and the path as the collection is awaiting approval and you get a failure of the task.

I think you should also do the check on version first to see if item exists, and if so then don't bother with attempting upload.

Tompage1994 commented:
> This flow doesn't quite work if you specify the version and the path as the collection is awaiting approval and you get a failure of the task.
> 
> I think you should also do the check on version first to see if item exists, and if so then don't bother with attempting upload.

`ansible -m  redhat_cop.ah_configuration.ah_collection -a "path=redhat_cop-ah_configuration-0.0.1-test.tar.gz namespace=redhat_cop name=ah_configuration version='0.0.1-test'" localhost -c local` fails with:

```
localhost | FAILED! => {
    "changed": false,
    "msg": "Could not find Collection redhat_cop.ah_configuration with_version 0.0.1-test"
}
```

despite the collection having uploaded and awaiting approval

Tompage1994 commented:
I wonder also if we should add the ability to auto-approve the collection in here too then implement the functionality of the ah_approval module here too.

sean-m-sullivan commented:
@Tompage1994 Reworked for workflow, also added an override, so it can just force a new version if the user wants by using the delete, and reworked the entire flow, to incorporate the approval and override.

Tompage1994 commented on plugins/module_utils/ah_module.py:
```diff
@@ -630,7 +630,8 @@ def wait_for_complete(self, task_url):
             self.fail_json(msg="Upload of collection failed: {0}".format(response["json"]["error"]["description"]))
         else:
-            self.exit_json(**self.json_output)
+            time.sleep(1)
```
I was going to complain about this, and perhaps we could do something such as provide an option for 'interval' and 'timeout' but I see this is exaclty what our friends in awx.awx do so I guess this is fine.

sean-m-sullivan replied:
I think I put in the sleep as between when it would upload, and the api would be populated it needed a second. Its for after the collection upload, not for waiting on it to be uploaded :)

Tompage1994 reviewed (commented):
Will give a proper review later when I get time

Tompage1994 commented on plugins/modules/ah_collection.py:
```diff
@@ -0,0 +1,169 @@
+    namespace: test_collection
+    name: test
+    version: 4.1.2
+    state: absent
```
This doesnt work. If you add in a version you get the following:

`fatal: [localhost]: FAILED! => {"changed": false, "msg": "Unable to approve at collections/redhat_cop/ah_configuration/versions/0.0.8-dev: Awaiting approval not found"}`

Tompage1994 replied:
i.e. not idempotent

sean-m-sullivan replied:
Fixed to be idempotent

Tompage1994 commented on plugins/modules/ah_collection.py:
```diff
@@ -0,0 +1,169 @@
+    name: test
+    version: 4.1.2
+    state: absent
+"""
```
Also 
```
- redhat_cop.ah_configuration.ah_collection:
        namespace: redhat_cop
        name: ah_configuration
        state: absent
        ah_host: https://localhost:8443
        ah_username: admin
        ah_password: password
        validate_certs: false
```

gives

`fatal: [localhost]: FAILED! => {"changed": false, "msg": "Could not find Collection redhat_cop.ah_configuration"}` when the collection isnt there

sean-m-sullivan replied:
Fixed for multiple scenarios with absent

Tompage1994 commented on plugins/modules/ah_collection.py:
```diff
@@ -0,0 +1,169 @@
+        - Approves a collection.
+        - Requires version to be set.
+      type: bool
+      default: true
```
It seems the behaviour I've seen is that it defaults to false?

Tompage1994 replied:
Actually, seeing that you need to set version too. That being said the version doesn't seem to really be used. We could work out the version based on whats passed in (the old module did that)

sean-m-sullivan replied:
Changed behavior to just approve after upload if its set to auto approve.

Tompage1994 reviewed (changes requested):
Quite a few things and we are a fair way off accepting this

sean-m-sullivan commented:
Sorry, busy week, Updated things, I think I've addressed the issues you found. I don't have an approval based AH atm, if you could test that? thanks

Tompage1994 commented on plugins/modules/ah_collection.py:
```diff
@@ -0,0 +1,178 @@
+    argument_spec = dict(
+        namespace=dict(required=True),
+        name=dict(required=True),
+        path=dict(),
```
Could we have a properly thrown error when the path is not found. I currently get a nasty stack trace if providing a path which is incorrect

sean-m-sullivan replied:
Add a pathlib check as that is supposed to be the module to use in python 3.4, on state = present to check for a path.

Tompage1994 commented on plugins/modules/ah_collection.py:
```diff
@@ -0,0 +1,178 @@
+        collection_endpoint = "collections/{0}/{1}".format(namespace, name)
+
+    existing_item = module.get_endpoint(collection_endpoint, **{"return_none_on_404": True})
+    if path:
```
Also 

```
- redhat_cop.ah_configuration.ah_collection:
        namespace: redhat_cop
        name: ah_configuration
        version: 0.0.9-dev
        path: ../redhat_cop-ah_configuration-0.0.9-dev.tar.gz
        state: absent
        ah_host: https://localhost:8443
        ah_username: admin
        ah_password: password
        validate_certs: false
```
will upload the collection despite `state: absent`. Need to switch around this if statement here because if state is absent then things certainly shouldnt be created

sean-m-sullivan replied:
Moved state absent to be checked first and to exit module after its tasks.

Tompage1994 reviewed (changes requested):
Looks much better. One error-handling and one functional issue then all good.
